subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -131,27 +134,43 @@ public boolean isDependencyToAll(String className) {
      * further dependents, while resources are just data accumulated along the way. Recurses for classes that
      * are "publicly accessbile", i.e. classes that are not just used privately in a class.
      */
-    private void processDependentClasses(Set<String> visitedClasses, Set<String> privateResultClasses, Set<String> accessibleResultClasses, Set<GeneratedResource> resultResources, Iterable<String> privateDependentClasses, Iterable<String> accessibleDependentClasses) {
-        for (String d : privateDependentClasses) {
-            if (!visitedClasses.add(d)) {
-                continue;
-            }
-            privateResultClasses.add(d);
-            DependentsSet currentDependents = getDependents(d);
-            if (!currentDependents.isDependencyToAll()) {
-                resultResources.addAll(currentDependents.getDependentResources());
-            }
-        }
-
-        for (String d : accessibleDependentClasses) {
-            if (!visitedClasses.add(d)) {
-                continue;
+    private void processDependentClasses(Set<String> visitedClasses,
+                                         Set<String> privateResultClasses,
+                                         Set<String> accessibleResultClasses,
+                                         Set<GeneratedResource> resultResources,
+                                         Iterable<String> privateDependentClasses,
+                                         Iterable<String> accessibleDependentClasses) {
+        List<String> remainingAccessibleDependentClasses = Lists.newArrayList(accessibleDependentClasses);
```

Comment:
I think this first can be extracted into a properly named method to make clear what this part of the algorithm is doing. For me it's hard to distinguish what the two loops are doing, how they are related to each other.